pywemo/subscribe.py

```diff
@@ -35,32 +73,124 @@ def get_ip_address(host='1.2.3.4'):
             for propnode in doc.findall('./{0}property'.format(NS)):
                 for property_ in list(propnode):
                     text = property_.text
                     outer.event(device, property_.tag, text)
 
-        self.send_response(200)
-        self.send_header('Content-Type', 'text/html')
-        self.send_header('Content-Length', len(SUCCESS))
+        self._send_response(200, RESPONSE_SUCCESS)
+
+    def do_GET(self):  # pylint: disable=invalid-name
+        """Handle GET requests for a Virtual WeMo device."""
+        if self.path.endswith("/setup.xml"):
+            self._send_response(
+                200, VIRTUAL_SETUP_XML, content_type="text/xml"
+            )
+        else:
+            self._send_response(404, RESPONSE_NOT_FOUND)
+
+    def do_POST(self):  # pylint: disable=invalid-name
+        """Handle POST requests for a Virtual WeMo device."""
+        if self.path.endswith("/upnp/control/basicevent1"):
+            sender_ip, _ = self.client_address
+            outer = self.server.outer
+            device = outer.devices.get(sender_ip)
+            if device is None:
+                LOG.warning(
+                    'Received event for unregistered device %s', sender_ip
+                )
+            else:
+                doc = self._get_xml_from_http_body()
+                binary_state = doc.find('.//BinaryState')
+                if binary_state is not None:
+                    text = binary_state.text
+                    outer.event(device, EVENT_TYPE_LONG_PRESS, text)
+            self._send_response(200, RESPONSE_SUCCESS)
+        else:
+            self._send_response(404, RESPONSE_NOT_FOUND)
+
+    def do_SUBSCRIBE(self):  # pylint: disable=invalid-name
+        """Handle SUBSCRIBE requests for a Virtual WeMo device."""
+        if self.path.endswith("/upnp/event/basicevent1"):
+            self.send_response(200)
+            self.send_header("CONTENT-LENGTH", "0")
+            self.send_header("TIMEOUT", "Second-1801")
+            self.send_header("SID", "uuid:virtual-device-uuid")
+            self.send_header('Connection', 'close')
+            self.end_headers()
+        else:
+            self._send_response(404, RESPONSE_NOT_FOUND)
+
+    def _send_response(self, code, body, *, content_type="text/html"):
+        self.send_response(code)
+        self.send_header('Content-Type', content_type)
+        self.send_header('Content-Length', len(body))
         self.send_header('Connection', 'close')
         self.end_headers()
-        self.wfile.write(SUCCESS.encode("UTF-8"))
+        if body:
+            self.wfile.write(body.encode("UTF-8"))
+
+    def _get_xml_from_http_body(self):
+        """Build the cElementTree root from the body of the http request."""
+        content_len = int(self.headers.get('content-length', 0))
+        data = self.rfile.read(content_len)
+        # trim garbage from end, if any
+        data = data.decode("UTF-8").strip()
```

Comment:
The previous version used `data.decode("UTF-8").split('\n\n')[0]` here.  Using strip is potentially different, but I think strip is the correct thing to do here.  It also removes any whitespace at the front, but I think we'd want to, if there was any there anyway.

Reply:
Wemo uses `\r\n\r\n` at the end of the POST messages, that's why I switched to `.strip()` here. I also think it's okay to remove whitespace at the front, but I could use `.rstrip()` if you'd prefer.

Reply:
I think keep strip